src/devices/MAX31856/MAX31856.cs

```diff
@@ -0,0 +1,209 @@
+        {
+            spiOutputData = WriteRead(Register.READ_CR0, 16);
+            double tempRaw = (((spiOutputData[13] & 0x7F) << 16) + (spiOutputData[14] << 8) + (spiOutputData[15]));
+            double temperature = tempRaw / 4096; // temperature in degrees C
+            if ((spiOutputData[13] & 0x80) == 0x80) // checks if the temp is negative
+            {
+                temperature = -temperature;
+            }
+
+            // Throws exceptions if the device is not found or if the thermocouple is not connected
+            if (spiOutputData[1] != (byte)Register.ONESHOT_FAULT_SETTING)
+            {
+                throw new IOException("Device not found");
+            }
+            else if (((spiOutputData[16] & (byte)Register.ERROR_OPEN) == (byte)Register.ERROR_OPEN))
+            {
+                throw new IOException("Thermocouple is not connected");
+            }
+
+            var temperatureOut = Temperature.FromDegreesCelsius(temperature); // Converts temperature to type Temperature struct and establishes it as Degrees C for its initial units
+
+            return temperatureOut;
+        }
+
+        /// <summary>
+        /// Converts Cold Junction Temperature Reading
+        /// </summary>
+        /// <remarks>
+        /// Takes the spi data as an input and outputs the Cold Junction Temperature Reading
+        /// </remarks>
+        /// <param name="spiOutputData">Spidata read from the device as 16 bytes</param>
+        private double ConvertspiOutputDataTempColdJunction(byte[] spiOutputData)
+        {
+            var tempRaw = (((spiOutputData[11] & 0x7F) << 8) + spiOutputData[12]);
+            double temperatureColdJunction = tempRaw / 256; // convert decimal ouput to temperature using a constant from the Technical Manual
+            if ((spiOutputData[11] & 0x80) == 0x80) // checks if the temp is negative
+            {
+                temperatureColdJunction = -temperatureColdJunction;
+            }
+
+            return temperatureColdJunction;
+        }
+
+        #endregion
+
+        #region read and write operations
+
+        /// <summary>
+        /// Writes the Data to the Spi Device
+        /// </summary>
+        /// <remarks>
+        /// Takes the data input byte and writes it to the spi device
+        /// </remarks>
+        /// <param name="data">Data to write to the device</param>
+        private void Write(Span<byte> data) => _spiDevice.Write(data);
+
+        /// <summary>
+        /// Full Duplex Read of the Data on the Device
+        /// </summary>
+        /// <remarks>
+        /// Writes the read address of the register and outputs a byte list of the length provided
+        /// </remarks>
+        /// /// <param name="register">Register location to write to which starts the device reading</param>
+        /// /// <param name="readbytesize">Number of bytes being read</param>
+        private byte[] WriteRead(Register register, int readbytesize)
```

Comment:
optional: consider changing this so that you do not need to allocate an output buffer (i.e. take Span as input instead of int and stackalloc at the callsite)

Reply:
Sorry this is too advanced for me :/

Reply:
no worries. in here we do not require having TryGetTemperature method, we can just to a single GetTemperature() or even property `Temperature`. TryGetTemperature method only makes sense if device is error prone (i.e. 1-wire thermomers are not working reliably as they are very time sensitive which we can't do correctly because OS can interrupt our process which causes error in transmission. In cases like that throwing and having to catch is a frequent operation and try .. catching is slower than not throwing at all. Easy way to explain is i.e.:

```csharp
boo TryGetTemperature(out Temperature temperature)
{
   temperature = ReadTemperatureFromRegister();
   if (DeviceStatuseShowsError())
   {
      temperature = default(Temperature);
      return false;
   }
  return true;
}
```

rather than throwing and try catching

Reply:
(as I said, in this case in this device errors are not expected so not having Try.. method is acceptable)

Reply:
I just hve the get temperature to simplify it. I agree with you that it was overly complicated for a simple device.